pkg/policy/api/rule.go

```diff
@@ -0,0 +1,259 @@
+// Copyright 2016-2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"github.com/cilium/cilium/pkg/labels"
+)
+
+// Rule is a policy rule which must be applied to all endpoints which match the
+// labels contained in the endpointSelector
+//
+// Each rule is split into an ingress section which contains all rules
+// applicable at ingress, and an egress section applicable at egress. For rule
+// types such as `L4Rule` and `CIDR` which can be applied at both ingress and
+// egress, both ingress and egress side have to either specifically allow the
+// connection or one side has to be omitted.
+//
+// Either ingress, egress, or both can be provided. If both ingress and egress
+// are omitted, the rule has no effect.
+type Rule struct {
+	// EndpointSelector selects all endpoints which should be subject to
+	// this rule. Cannot be empty.
+	EndpointSelector EndpointSelector `json:"endpointSelector"`
+
+	// Ingress is a list of IngressRule which are enforced at ingress.
+	// If omitted or empty, this rule does not apply at ingress.
+	//
+	// +optional
+	Ingress []IngressRule `json:"ingress,omitempty"`
+
+	// Egress is a list of EgressRule which are enforced at egress.
+	// If omitted or empty, this rule does not apply at egress.
+	//
+	// +optional
+	Egress []EgressRule `json:"egress,omitempty"`
+
+	// Labels is a list of optional strings which can be used to
+	// re-identify the rule or to store metadata. It is possible to lookup
+	// or delete strings based on labels. Labels are not required to be
+	// unique, multiple rules can have overlapping or identical labels.
+	//
+	// +optional
+	Labels labels.LabelArray `json:"labels,omitempty"`
+
+	// Description is a free form string, it can be used by the creator of
+	// the rule to store human readable explanation of the purpose of this
+	// rule. Rules cannot be identified by comment.
+	//
+	// +optional
+	Description string `json:"description,omitempty"`
+}
+
+// IngressRule contains all rule types which can be applied at ingress,
+// i.e. network traffic that originates outside of the endpoint and
+// is entering the endpoint selected by the endpointSelector.
+//
+// - All members of this structure are optional. If omitted or empty, the
+//   member will have no effect on the rule.
+// - All members of this structure are evaluated independently, i.e. L4 ports
```

Comment:
Correct. Two reasons:
- The BPF datapath is currently not capable of enforcing this. L3 endpoints are whitelisted via BPF map. The per endpoint L4 map is another BPF map. Referring to a 2nd BPF map from a 1st BPF map value was only just recently added so we will start supporting this with more recent kernels.
- k8s NetworkPolicy also defines this separate notion and we need to able to map k8s NetworkPolicy to our policy.

The idea is to add a field later on which introduces L4 policies which depend on a specific L3 rule. This could be done by adding a new `ToPorts` inside `FromEndpoints` or to add a flag to `ToPorts` which allows to tie the `ToPorts` to `FromEndpoints` and `FromCIDR` within the same IngressRule.